File: frontend/src/app/core/services/equipment.service.ts
import { Injectable } from '@angular/core';
import { from, Observable } from 'rxjs';
import { EquipmentDto, CreateEquipmentRequest, CreateMaintenanceTaskRequest, MaintenanceTaskDto } from '../models/equipment.model';
import { db, EquipmentRecord, MaintenanceTaskRecord } from '../db/bastion-db';

function todayStr(): string { return new Date().toISOString().split('T')[0]; }
function soonStr(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() + days);
  return d.toISOString().split('T')[0];
}

function computeNextDue(task: MaintenanceTaskRecord, equip: EquipmentRecord): string | null {
  if (!task.intervalDays) return null;
  const base = task.lastCompletedAt ?? equip.purchaseDate;
  const d = new Date(base);
  d.setDate(d.getDate() + task.intervalDays);
  return d.toISOString().split('T')[0];
}

function toTaskDto(task: MaintenanceTaskRecord, equip: EquipmentRecord): MaintenanceTaskDto {
  const today = todayStr();
  const soon14 = soonStr(14);
  const nextDueAt = computeNextDue(task, equip);
  const isOverdue = task.intervalDays != null && nextDueAt != null && nextDueAt < today;
  const isDueSoon = !isOverdue && nextDueAt != null && nextDueAt <= soon14;
  return {
    id: task.id,
    equipmentId: task.equipmentId,
    description: task.description,
    intervalDays: task.intervalDays,
    lastCompletedAt: task.lastCompletedAt ?? undefined,
    nextDueAt,
    isOverdue,
    isDueSoon
  };
}

@Injectable({ providedIn: 'root' })
export class EquipmentService {
  getAll(): Observable<EquipmentDto[]> {
    return from(this.getAllAsync());
  }

  create(req: CreateEquipmentRequest): Observable<EquipmentDto> {
    return from(this.createAsync(req));
  }

  delete(id: string): Observable<void> {
    return from(
      db.transaction('rw', db.equipment, db.maintenanceTasks, async () => {
        await db.maintenanceTasks.where('equipmentId').equals(id).delete();
        await db.equipment.delete(id);
      }).then(() => undefined as void)
    );
  }

  addTask(equipmentId: string, req: CreateMaintenanceTaskRequest): Observable<MaintenanceTaskDto> {
    return from(this.addTaskAsync(equipmentId, req));
  }

  completeTask(equipmentId: string, taskId: string): Observable<MaintenanceTaskDto> {
    return from(this.completeTaskAsync(equipmentId, taskId));
  }

  deleteTask(_equipmentId: string, taskId: string): Observable<void> {
    return from(db.maintenanceTasks.delete(taskId).then(() => undefined as void));
  }

  private async getAllAsync(): Promise<EquipmentDto[]> {
    const [equipment, tasks] = await Promise.all([
      db.equipment.toArray(),
      db.maintenanceTasks.toArray()
    ]);
    return equipment.map(e => ({
      id: e.id,
      householdId: 'local',
      name: e.name,
      category: e.category,
      purchaseDate: e.purchaseDate,
      tasks: tasks.filter(t => t.equipmentId === e.id).map(t => toTaskDto(t, e))
    }));
  }

  private async createAsync(req: CreateEquipmentRequest): Promise<EquipmentDto> {
    const id = crypto.randomUUID();
    const record: EquipmentRecord = { id, name: req.name, category: req.category, purchaseDate: req.purchaseDate };
    await db.equipment.add(record);
    return { id, householdId: 'local', name: req.name, category: req.category, purchaseDate: req.purchaseDate, tasks: [] };
  }

  private async addTaskAsync(equipmentId: string, req: CreateMaintenanceTaskRequest): Promise<MaintenanceTaskDto> {
    const equip = await db.equipment.get(equipmentId);
    if (!equip) throw new Error(`Equipment ${equipmentId} not found`);
    const task: MaintenanceTaskRecord = {
      id: crypto.randomUUID(),
      equipmentId,
      description: req.description,
      intervalDays: req.intervalDays,
      lastCompletedAt: null
    };
    await db.maintenanceTasks.add(task);
    return toTaskDto(task, equip);
  }

  private async completeTaskAsync(equipmentId: string, taskId: string): Promise<MaintenanceTaskDto> {
    const [task, equip] = await Promise.all([
      db.maintenanceTasks.get(taskId),
      db.equipment.get(equipmentId)
    ]);
    if (!task || !equip) throw new Error('Not found');
    const updated: MaintenanceTaskRecord = { ...task, lastCompletedAt: todayStr() };
    await db.maintenanceTasks.put(updated);
    return toTaskDto(updated, equip);
  }
}

File: src/Bastion.Application/Equipment/EquipmentService.cs
using Bastion.Application.Equipment.Dtos;
using Bastion.Application.Dashboard;
using Bastion.Domain.Aggregates.Equipment;

namespace Bastion.Application.Equipment;

public interface IEquipmentService
{
    Task<IReadOnlyList<EquipmentDto>> GetByHouseholdAsync(Guid householdId, CancellationToken ct = default);
    Task<EquipmentDto?> GetByIdAsync(Guid id, CancellationToken ct = default);
    Task<EquipmentDto> CreateAsync(Guid householdId, CreateEquipmentRequest request, CancellationToken ct = default);
    Task<EquipmentDto?> UpdateAsync(Guid id, UpdateEquipmentRequest request, CancellationToken ct = default);
    Task<bool> DeleteAsync(Guid id, CancellationToken ct = default);
    Task<MaintenanceTaskDto?> AddTaskAsync(Guid equipmentId, CreateMaintenanceTaskRequest request, CancellationToken ct = default);
    Task<MaintenanceTaskDto?> UpdateTaskAsync(Guid equipmentId, Guid taskId, UpdateMaintenanceTaskRequest request, CancellationToken ct = default);
    Task<MaintenanceTaskDto?> CompleteTaskAsync(Guid equipmentId, Guid taskId, DateOnly completedAt, CancellationToken ct = default);
    Task<bool> DeleteTaskAsync(Guid equipmentId, Guid taskId, CancellationToken ct = default);
}

public class EquipmentService(IEquipmentRepository repository) : IEquipmentService
{
    private static readonly DateOnly Today = DateOnly.FromDateTime(DateTime.UtcNow);

    public async Task<IReadOnlyList<EquipmentDto>> GetByHouseholdAsync(Guid householdId, CancellationToken ct = default)
    {
        var list = await repository.GetByHouseholdAsync(householdId, ct);
        return list.Select(ToDto).ToList();
    }

    public async Task<EquipmentDto?> GetByIdAsync(Guid id, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(id, ct);
        return eq is null ? null : ToDto(eq);
    }

    public async Task<EquipmentDto> CreateAsync(Guid householdId, CreateEquipmentRequest request, CancellationToken ct = default)
    {
        var eq = Domain.Aggregates.Equipment.Equipment.Create(
            request.Name, request.Category, request.PurchaseDate, householdId);
        await repository.AddAsync(eq, ct);
        return ToDto(eq);
    }

    public async Task<EquipmentDto?> UpdateAsync(Guid id, UpdateEquipmentRequest request, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(id, ct);
        if (eq is null) return null;
        eq.Update(request.Name, request.Category, request.PurchaseDate);
        await repository.SaveAsync(ct);
        return ToDto(eq);
    }

    public async Task<bool> DeleteAsync(Guid id, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(id, ct);
        if (eq is null) return false;
        repository.Remove(eq);
        await repository.SaveAsync(ct);
        return true;
    }

    public async Task<MaintenanceTaskDto?> AddTaskAsync(Guid equipmentId, CreateMaintenanceTaskRequest request, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(equipmentId, ct);
        if (eq is null) return null;
        var task = eq.AddTask(request.Description, request.IntervalDays);
        repository.AddTask(task);
        await repository.SaveAsync(ct);
        return ToTaskDto(task);
    }

    public async Task<MaintenanceTaskDto?> UpdateTaskAsync(Guid equipmentId, Guid taskId, UpdateMaintenanceTaskRequest request, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(equipmentId, ct);
        var task = eq?.Tasks.FirstOrDefault(t => t.Id == taskId);
        if (task is null) return null;
        task.Update(request.Description, request.IntervalDays);
        await repository.SaveAsync(ct);
        return ToTaskDto(task);
    }

    public async Task<MaintenanceTaskDto?> CompleteTaskAsync(Guid equipmentId, Guid taskId, DateOnly completedAt, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(equipmentId, ct);
        var task = eq?.Tasks.FirstOrDefault(t => t.Id == taskId);
        if (task is null) return null;
        task.Complete(completedAt);
        await repository.SaveAsync(ct);
        return ToTaskDto(task);
    }

    public async Task<bool> DeleteTaskAsync(Guid equipmentId, Guid taskId, CancellationToken ct = default)
    {
        var eq = await repository.GetByIdAsync(equipmentId, ct);
        if (eq is null) return false;
        var removed = eq.RemoveTask(taskId);
        if (!removed) return false;
        await repository.SaveAsync(ct);
        return true;
    }

    private MaintenanceTaskDto ToTaskDto(MaintenanceTask t) =>
        new(t.Id, t.EquipmentId, t.Description, t.IntervalDays,
            t.LastCompletedAt, t.NextDueAt,
            t.IsOverdue(Today), t.IsDueSoon(Today));

    private EquipmentDto ToDto(Domain.Aggregates.Equipment.Equipment e) =>
        new(e.Id, e.HouseholdId, e.Name, e.Category, e.PurchaseDate,
            e.Tasks.Select(ToTaskDto).ToList());
}
